### PR TITLE
Add reflog support

### PR DESCRIPTION
**Pull request by Soph**
Add support for reading and writing Git reflogs across the storage layer.

- `plumbing/format/reflog` - Encoder/decoder for the reflog file format. Handles the `<old> <new> <name> <<email>> <timestamp> <tz>\t<message>\n` line format, with proper error reporting for malformed entries and message normalization matching Git's behavior
- `plumbing/storer` - New `ReflogStorer` interface (`Reflog`, `AppendReflog`, `DeleteReflog`), kept separate from `storage.Storer` so reflog support is opt-in.
- `storage/filesystem` - `ReflogStorage` backed by `.git/logs/` files.
- `storage/memory` - In-memory `ReflogStorer` implementation.
- `storage/transactional` - Transactional reflog with proper isolation - tracks appends and deletes independently and flushes on `Commit()`. Conditionally created only when both base and temporal storers support reflogs.

Design decisions

- `ReflogStorer` is not part of `storage.Storer` - callers opt in via type assertion, so existing code is unaffected.
- `Reflog()` returns entries oldest-first (file order), matching the on-disk format.
- Message normalization on encode (collapse whitespace, strip newlines) matches Git's behavior and prevents format corruption.
- Empty messages omit the TAB separator, matching Git.

Test plan

- Unit tests for encode/decode round-trips, edge cases (empty messages, long lines, malformed timestamps)
- Transactional storage test suite covering isolation, commit, and concurrent base/temporal operations
- Verified output matches git reflog on real repositories

### CODE REVIEW

File: plumbing/format/reflog/doc.go [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// Package reflog implements encoding and decoding of git reflog entries.
+//
+// Git stores reflog entries in .git/logs/<ref> files, one entry per line.
+// Each line has the format:
+//
+//	<old-hash> <new-hash> <name> <<email>> <unix-timestamp> <timezone>\t<message>\n
```
[the remark this answers is not in the document]
Comment (Soph) [reply plumbing/format/reflog/doc.go:6]: this is correct, the `<>` are both placeholders for values but `email` has actual `<>` wrapping too in a reflog. See for example `.git/logs/HEAD`

File: plumbing/format/reflog/reflog.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+		}
+
+		if err == io.EOF {
+			return entries, nil
+		}
+	}
+}
+
+// decodeLine parses a single reflog line.
+// Format: <old-hash> <new-hash> <name> <<email>> <unix-timestamp> <timezone>\t<message>
```
[the remark this answers is not in the document]
Comment (Soph) [reply plumbing/format/reflog/reflog.go:89]: this is correct, the `<>` are both placeholders for values but `email` has actual `<>` wrapping too in a reflog. See for example `.git/logs/HEAD`

File: plumbing/format/reflog/reflog_test.go [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+	t.Parallel()
+
+	message := strings.Repeat("a", 70*1024)
+	line := "0000000000000000000000000000000000000000 aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa Author <a@b.com> 1234567890 +0000\t" + message + "\n"
+
+	entries, err := Decode(strings.NewReader(line))
+	require.NoError(t, err)
+	require.Len(t, entries, 1)
+	assert.Equal(t, message, entries[0].Message)
+}
```
Comment (pjbgf) [inline plumbing/format/reflog/reflog_test.go:295]: Let's add a Fuzz test for it as well.

Something like the one [for worktree](https://github.com/go-git/go-git/blob/5b5182acfe97427c3245f53e71afc114e79e4fc6/x/plumbing/worktree/worktree_test.go#L1244), whereby we add some valid and invalid data to the fuzzing corpus.

File: plumbing/format/reflog/doc.go (hunk 1 above)
Comment (pjbgf) [inline plumbing/format/reflog/doc.go:6]: For clarity:
```suggestion
//	<old-hash> <new-hash> <committer> <<email>> <unix-timestamp> <timezone>\t<message>\n
```

File: plumbing/format/reflog/reflog.go [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+	// OldHash is the hash the reference pointed to before the change.
+	OldHash plumbing.Hash
+	// NewHash is the hash the reference points to after the change.
+	NewHash plumbing.Hash
+	// Name is the name of the person who made the change.
+	Name string
+	// Email is the email of the person who made the change.
+	Email string
+	// When is the timestamp of the change.
+	When time.Time
```
Comment (pjbgf) [inline plumbing/format/reflog/reflog.go]: Align with the approach in `Commit`, encapsulating the fields `Name`, `Email` and `When` into a single object:

```suggestion
	// Committer holds the signature for the entry, including name, email and when it was created.
	Committer object.Signature
```

File: plumbing/format/reflog/reflog.go [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+	Message string
+}
+
+// Decode reads all reflog entries from the reader.
+// Entries are returned in file order (oldest first).
+func Decode(r io.Reader) ([]*Entry, error) {
+	var entries []*Entry
+	reader := bufio.NewReader(r)
+	for {
+		line, err := reader.ReadBytes('\n')
```
Comment (pjbgf) [inline plumbing/format/reflog/reflog.go]: I suspect that if `r` is `nil`, this will panic. Please add a test to that effect and a `nil` check if that's the case.

File: plumbing/format/reflog/reflog_test.go [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+func TestDecodeInvalidNewHash(t *testing.T) {
+	t.Parallel()
+
+	line := "0000000000000000000000000000000000000000 not-a-valid-hash Author <a@b.com> 1234567890 +0000\tcommit: test\n"
+	_, err := Decode(strings.NewReader(line))
+	require.Error(t, err)
+	assert.Contains(t, err.Error(), "invalid new hash")
+}
+
```
Comment (pjbgf) [inline plumbing/format/reflog/reflog_test.go:259]: Please add a test whereby both commits are 64 characters to represent `SHA256`.

File: plumbing/format/reflog/reflog.go [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+func normalizeMessage(msg string) string {
+	msg = strings.ReplaceAll(msg, "\n", " ")
+	msg = strings.ReplaceAll(msg, "\r", " ")
+	fields := strings.Fields(msg)
+	return strings.Join(fields, " ")
+}
+
+// Encode writes a single reflog entry to the writer.
+func Encode(w io.Writer, e *Entry) error {
+	_, offset := e.When.Zone()
```
Comment (pjbgf) [inline plumbing/format/reflog/reflog.go]: Please add `nil` checks for `w` and `e`.

File: plumbing/format/reflog/reflog_test.go (hunk 3 above)
Comment (Soph) [reply plumbing/format/reflog/reflog_test.go:295]: added.

File: plumbing/format/reflog/doc.go (hunk 1 above)
Comment (Soph) [reply plumbing/format/reflog/doc.go:6]: changed

File: plumbing/format/reflog/reflog.go (hunk 4 above)
Comment (Soph) [reply plumbing/format/reflog/reflog.go]: Couldn't use `object.Signature` because that would have created a dependency cycle: `reflog -> object -> storer -> reflog` but I think this is now fine too.

File: plumbing/format/reflog/reflog.go (hunk 5 above)
Comment (Soph) [reply plumbing/format/reflog/reflog.go]: added.

File: plumbing/format/reflog/reflog_test.go (hunk 6 above)
Comment (Soph) [reply plumbing/format/reflog/reflog_test.go:259]: added.

File: plumbing/format/reflog/reflog.go (hunk 7 above)
Comment (Soph) [reply plumbing/format/reflog/reflog.go]: added

File: plumbing/format/reflog/reflog.go (hunk 4 above)
Comment (pjbgf) [reply plumbing/format/reflog/reflog.go]: As per off-line chat, let's remove this - the idea was to rely on existing patterns. Let's keep as is instead.

File: plumbing/format/reflog/reflog.go [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+	NewHash plumbing.Hash
+	// Committer holds the signature for the entry, including name, email and when it was created.
+	Committer Signature
+	// Message describes the action that caused the change (e.g. "commit: Add feature").
+	Message string
+}
+
+// Decode reads all reflog entries from the reader.
+// Entries are returned in file order (oldest first).
+func Decode(r io.Reader) ([]*Entry, error) {
```
Comment (onee-only) [inline plumbing/format/reflog/reflog.go]: Maybe we can make it return an iterator? Some people might not want to load the whole entries in their memory.

File: plumbing/format/reflog/reflog.go [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+// normalizeMessage normalizes a reflog message the same way Git does:
+// collapse consecutive whitespace to a single space, strip leading/trailing
+// whitespace, and remove newlines.
+func normalizeMessage(msg string) string {
+	msg = strings.ReplaceAll(msg, "\n", " ")
+	msg = strings.ReplaceAll(msg, "\r", " ")
+	fields := strings.Fields(msg)
+	return strings.Join(fields, " ")
+}
```
Comment (onee-only) [inline plumbing/format/reflog/reflog.go:193]: Would you kindly share a reference of this implementation?

Comment (Soph) [reply plumbing/format/reflog/reflog.go:193]: Good point. Added a comment with the source.

File: plumbing/format/reflog/reflog.go (hunk 8 above)
Comment (Soph) [reply plumbing/format/reflog/reflog.go]: Added a iterator and kept the method to return all.

File: plumbing/format/reflog/reflog.go [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+	e.OldHash = plumbing.NewHash(string(line[:spaceIdx]))
+	line = line[spaceIdx+1:]
+
+	// Parse new hash (up to next space)
+	spaceIdx = bytes.IndexByte(line, ' ')
+	if spaceIdx == -1 {
+		return nil, fmt.Errorf("expected space after new hash")
+	}
+	e.NewHash = plumbing.NewHash(string(line[:spaceIdx]))
+	line = line[spaceIdx+1:]
```
[the remark this answers is not in the document]
Comment (pjbgf) [reply plumbing/format/reflog/reflog.go:115]: This would be a valid feedback if not preceded by `plumbing.IsHash(newHashStr)`.

**Review by pjbgf — APPROVED** [review]
@Soph thanks for working on this. 🙇
